tools/c7n_oci/c7n_oci/session.py

```diff
@@ -0,0 +1,103 @@
+# Copyright The Cloud Custodian Authors.
+# SPDX-License-Identifier: Apache-2.0
+
+import importlib
+import logging
+import os
+
+import oci
+
+from c7n_oci.constants import (
+    ENV_FINGERPRINT,
+    ENV_USER,
+    ENV_KEY_FILE,
+    ENV_REGION,
+    ENV_TENANCY,
+    DEFAULT_PROFILE,
+)
+
+log = logging.getLogger("custodian.oci.session")
+
+
+class Session:
+    def __init__(self, config_file_path=None, profile_name=DEFAULT_PROFILE):
+        self.config = None
+        self.authenticated = False
+        self.config_file_path = config_file_path
+        self.profile_name = profile_name
+
+    def initialize_session(self):
+        if self.authenticated:
+            return
+        if os.environ.get(ENV_FINGERPRINT) is not None:
+            self.config = {
+                "fingerprint": os.environ.get(ENV_FINGERPRINT),
+                "key_file": os.environ.get(ENV_KEY_FILE),
+                "region": os.environ.get(ENV_REGION),
+                "tenancy": os.environ.get(ENV_TENANCY),
+                "user": os.environ.get(ENV_USER),
+            }
+        elif self.profile_name is not None and self.config_file_path is None:
+            self.config = oci.config.from_file(profile_name=self.profile_name)
+        elif self.profile_name is None and self.config_file_path is not None:
+            self.config = oci.config.from_file(self.config_file_path)
+        elif self.profile_name and self.config_file_path:
+            self.config = oci.config.from_file(self.config_file_path, self.profile_name)
+        else:
+            self.config = oci.config.from_file()
+
+        self.config["additional_user_agent"] = (
+            f'Oracle-CloudCustodian {self.config["additional_user_agent"]}'
+            if self.config.get("additional_user_agent")
+            else "Oracle-CloudCustodian"
+        )
+
+        # The next statements are just to verify that the config is working
+        try:
+            identity_client = oci.identity.IdentityClient(self.config)
+
+            response = identity_client.get_user(self.config.get("user"))
+            self.authenticated = True
+            log.info(f"Successfully authenticated user [{response.data.name}]")
+        except Exception as e:
+            log.error("Failed to authenticate.\nMessage: {}".format(e))
+
+    def get_oci_config(self):
+        return self.config
+
+    def client(self, client):
+        self.initialize_session()
```

Comment:
session initialization is once per process and is currently handled during provider initialization. testing may need a helper but we shouldn't have to do this for each client. this also applies to get_monitoring_client, and get_work_request_client.

Reply:
@kapilt session.initialize_session() has the 'authenticated' flag which will be set when the user gets authenticated for the first time. From the next time, we will return from the session.initialize_session() function without doing anything is the user is authenticated( if the 'authenticated' flag is set). The same mechanism is followed in the other providers too. Correct us if we are missing anything.

Reply:
ic, I'm okay with rolling with this as is for now.

for the other providers, its intended as one time across session instances typically via configuration on options, on the additional api calls to get account id typically. then the per session instance api calls are limited to authentication credential retrieval. looking more closely it does look like the azure provider served as an inspiration, and I do see something similar there wrt to per instance limited to auth. the difference in the oci provider is the the api call we're issuing in session initialization, is not related to retrieving auth but verifying auth, which means its extraneous to do it more than once per process lifetime, and its currently scoped to every session instance.